File: projects/agentic-automation-framework/src/agent/semantic_memory.js
import fs from "fs";
import path from "path";

const REPORT_DIR = path.resolve("reports");
const EMB_FILE = path.join(REPORT_DIR, "agent-memory-embeddings.json");

function ensure() {
  fs.mkdirSync(REPORT_DIR, { recursive: true });
}

function dot(a, b) {
  let s = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) s += a[i] * b[i];
  return s;
}

function norm(a) {
  return Math.sqrt(dot(a, a)) || 1;
}

function cosine(a, b) {
  return dot(a, b) / (norm(a) * norm(b));
}

export function loadEmbMemory() {
  ensure();
  if (!fs.existsSync(EMB_FILE)) return [];
  try {
    return JSON.parse(fs.readFileSync(EMB_FILE, "utf-8"));
  } catch {
    return [];
  }
}

export function saveEmbMemory(items) {
  ensure();
  fs.writeFileSync(EMB_FILE, JSON.stringify(items.slice(-200), null, 2));
}

export async function embedText(text) {
  const key = process.env.OPENAI_API_KEY;
  if (!key) return null;

  const model = process.env.OPENAI_EMBED_MODEL || "text-embedding-3-small";
  const base = (process.env.OPENAI_BASE_URL || "https://api.openai.com/v1").replace(/\/$/, "");
  const res = await fetch(`${base}/embeddings`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${key}`
    },
    body: JSON.stringify({ model, input: text })
  });

  if (!res.ok) return null;
  const data = await res.json();
  return data?.data?.[0]?.embedding || null;
}

export async function rememberFailureSemantic(item) {
  const vec = await embedText(`${item.goal}\n${item.error}`);
  if (!vec) return false;
  const mem = loadEmbMemory();
  mem.push({ ...item, embedding: vec });
  saveEmbMemory(mem);
  return true;
}

export async function retrieveRelevantSemantic(goal, limit = 5) {
  const q = await embedText(goal);
  if (!q) return [];
  const mem = loadEmbMemory();
  const scored = mem
    .map((m) => ({ ...m, score: cosine(q, m.embedding || []) }))
    .filter((m) => Number.isFinite(m.score))
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
  return scored;
}

File: projects/agentic-automation-framework-python/agentic/agent/semantic_memory.py
import json
from pathlib import Path
import requests
from agentic.config import CONFIG

REPORTS = Path("reports")
SEM = REPORTS / "agent-memory-embeddings.json"


def _ensure():
    REPORTS.mkdir(exist_ok=True)


def _load():
    _ensure()
    if not SEM.exists():
        return []
    try:
        return json.loads(SEM.read_text())
    except Exception:
        return []


def _save(items):
    SEM.write_text(json.dumps(items[-200:], indent=2))


def _dot(a, b):
    return sum(x * y for x, y in zip(a, b))


def _norm(a):
    return (_dot(a, a) ** 0.5) or 1.0


def _cos(a, b):
    return _dot(a, b) / (_norm(a) * _norm(b))


def embed_text(text: str):
    key = CONFIG["openai"]["key"]
    if not key:
        return None
    base = CONFIG["openai"]["base_url"].rstrip("/")
    model = CONFIG["openai"]["embed_model"]
    try:
        r = requests.post(
            f"{base}/embeddings",
            headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
            json={"model": model, "input": text},
            timeout=40,
        )
        if not r.ok:
            return None
        return r.json().get("data", [{}])[0].get("embedding")
    except Exception:
        return None


def remember_semantic(item: dict):
    vec = embed_text(f"{item.get('goal','')}\n{item.get('error','')}")
    if not vec:
        return False
    mem = _load()
    mem.append({**item, "embedding": vec})
    _save(mem)
    return True


def relevant_semantic(goal: str, limit=5):
    q = embed_text(goal)
    if not q:
        return []
    mem = _load()
    scored = []
    for m in mem:
        vec = m.get("embedding")
        if not vec:
            continue
        s = _cos(q, vec)
        x = dict(m)
        x["score"] = s
        scored.append(x)
    scored.sort(key=lambda x: x["score"], reverse=True)
    return scored[:limit]
